plugin/resolve/resolve.go

```diff
@@ -0,0 +1,123 @@
+package resolve
+
+import (
+	"context"
+	"strings"
+
+	"github.com/coredns/coredns/plugin"
+	"github.com/coredns/coredns/plugin/pkg/nonwriter"
+	"github.com/coredns/coredns/request"
+
+	"github.com/miekg/dns"
+)
+
+// Resolve performs CNAME target resolution on all CNAMEs in the response
+type Resolve struct {
+	Next    plugin.Handler
+	Zones   []string
+	DoCNAME bool
+	DoSRV   bool
+}
+
+// Name implements the Handler interface.
+func (c Resolve) Name() string { return name() }
+func name() string             { return "resolve" }
+
+// ServeDNS implements the plugin.Handle interface.
+func (c Resolve) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	state := request.Request{W: w, Req: r, Context: ctx}
+
+	zone := plugin.Zones(c.Zones).Matches(state.Name())
+	if zone == "" {
+		return plugin.NextOrFailure(c.Name(), c.Next, ctx, w, r)
+	}
+
+	if state.QType() == dns.TypeCNAME {
+		return plugin.NextOrFailure(c.Name(), c.Next, ctx, w, r)
+	}
+
+	// Run the query through the rest of the plugin chain using a non-writer
+	nw := nonwriter.New(w)
+	rcode, err := plugin.NextOrFailure(c.Name(), c.Next, ctx, nw, r)
+	if err != nil || !plugin.ClientWrite(rcode) {
+		return rcode, err
+	}
+
+	// Look at each answer and do lookups for any CNAME/SRV answers
+	for i := 0; i < len(nw.Msg.Answer); i++ {
```

Comment:
this should probably be done in parallel? (at some point??)

Reply:
Perhaps - but for CNAMES at least I think usually you only get one CNAME in a response.  Exceptions are corner case (i.e. wildcard queries).

For SRV, you can get multiples if you leave the service protocol prefixes off the query name.  But not sure how common that really is. 

So yeah - theoretically would be better to do them in parallel, but not sure how much practical benefit we'd see.